### DISCUSSION TITLE
Is caffeine supported in reactive programming

### DISCUSSION THREAD

**Discussion by omikhare** [ben-manes/caffeine#648] [Q&A] [answered]
Hi Ben, 

I am looking for caffeine use in spring web flux which support only non blocking APIs. Does caffeine provides either async or reactive template to support spring web flux.

Thank.

**Comment by ben-manes** [comment, marked as answer]
You can use `AsyncCache` and convert using `Mono.fromFuture` and `Mono.toFuture`. Spring Cache does not have async support (https://github.com/spring-projects/spring-framework/issues/17920), so you should use our apis directly.

**Comment by alex-arana** [comment]
You can integrate Caffeine caches with Spring WebFlux using [reactor-addons](https://github.com/reactor/reactor-addons) utility classes [CacheMono](https://projectreactor.io/docs/extra/snapshot/api/reactor/cache/CacheMono.html) and [CacheFlux](https://projectreactor.io/docs/extra/snapshot/api/reactor/cache/CacheFlux.html). However, both have been deprecated and have been slated for removal in [release 3.6.0](https://github.com/reactor/reactor-addons/issues/263).

I recently wrote a replacement (in Kotlin) for both of those helper classes using a couple of different strategies:
- A local caching strategy backed by an instance of `AsyncLoadingCache`
- A delegated approach which adapts a Spring Cache instance (ie. _Redis_) to an `AsyncCacheLoader` so that value loading occurs in a Future object thus addressing potential [cache stampeding](https://en.wikipedia.org/wiki/Cache_stampede) issues.

I had some existing APIs that needed to retro-fitted so would need to untangle all the relevant bits but happy to share if anyone is interested..
